Captcha 예측 기능을 포함한 FastAPI 백엔드 구축

RaonJena99 opened the pull request:
- CORS 및 세션 미들웨어 설정
- /captcha 엔드포인트로 캡차 문제 생성 및 저장
- /predict 엔드포인트로 캡차 이미지 추론 처리
- MLflow에서 등록된 PyTorch 모델 불러와 예측 수행
- HybridCNN 모델 클래스 정의
- image_tensor에 .to(device) 추가하여 모델과 디바이스 일치시킴

RaonJena99 committed:
프론트엔드,백엔드,ML모델 연결

- CORS 미들웨어로 프론트엔드와 백엔드 연결
- mlflow에 등록된 모델을 이용해서 예측한 값을 다시 프론트엔드로 전달

RaonJena99 committed:
CNN 이름 수정

- 기존 SimpleCNN에서 HybridCNN으로 이름 변경

RaonJena99 committed:
입력 텐서와 모델 간 디바이스 불일치 오류 해결 (MPS 환경 대응)

- 모델은 GPU(MPS)에 로딩되었지만, 입력 이미지 텐서는 CPU에 존재하여 RuntimeError 발생
- image_tensor에 .to(device) 추가하여 모델과 디바이스 일치시킴
- Mac + MPS 환경에서 발생하는 오류에 대한 대응

RaonJena99 committed:
이미지 전처리 개선 및 입력 크기 일관성 유지

- 글자 중심 정렬, padding 적용, 28x28 resize하여 모델 학습과 동일한 입력 포맷 유지
- resize된 이미지를 저장하여 디버깅 및 성능 분석에 용이하도록 구성

RaonJena99 commented:
이미지 전처리 개선 및 입력 크기 일관성 유지
- 글자 중심 정렬, padding 적용, 28x28 resize하여 모델 학습과 동일한 입력 포맷 유지
- resize된 이미지를 저장하여 디버깅 및 성능 분석에 용이하도록 구성

ljy2855 commented on src/mlruns/0/meta.yaml:
mlflow를 로컬에 두지 않고 remote로 접근해서 사용할 예정이라 해당 mlruns는 지워주세요!

ljy2855 commented on src/models/models.py:
마찬가지로 model도 mlflow로부터 load하여 사용할 예정이라, 해당 코드도 작동 확인되셧으면 지워주세요!

ljy2855 commented on src/routers/captcha.py:
```diff
@@ -1,15 +1,70 @@
 
 router = APIRouter()
 
+# 디바이스 설정
```
router폴더에는 로직 작성보다 endpoint 관리를 위한 파일이라 device 세팅은 models 모듈에 utils 파일에 작성해주세요

ljy2855 commented on src/routers/captcha.py:
```diff
@@ -1,15 +1,70 @@
 # 메모리 기반 문제 저장소 (임시용)
 captcha_store = {}
 
+# base64 이미지 데이터 중심 정렬 전처리
```
마찬가지로 이미지 전처리 로직도 utils에 작성해주세요

ljy2855 reviewed (changes requested):
어려우실 수 있으실텐데, 잘 해주셔서 감사합니다 :)

backend 서비스다보니 ml 모델에 대한 dependency만 제거해주시면 좋을 것 같아요!
[모델 서빙](https://github.com/ljy2855/stock_prediction/blob/main/model_serve/main.py) 이 파일을 한번 보시고 수정만 부탁드려요!

RaonJena99 committed:
백엔드에서 ML 의존성 제거 및 추론 기능을 원격 ML 서비스로 위임

- 로컬 ML 모델 로딩 및 MLflow 의존성 제거
- httpx를 통해 원격 ML 추론 API로 예측 로직 위임
- 이미지 전처리 로직을 utils/image_processing.py로 분리

RaonJena99 commented:
수정을 했는데, 제대로 작동하는지는 확인하지 못했습니다.

ljy2855 commented on src/routers/captcha.py:
```diff
@@ -1,108 +1,56 @@
-model = mlflow.pytorch.load_model("models:/HybridCNN@production")
-model.to(device)
-model.eval()
+# 원격 ML 서비스 URL
```
이미 main에서 env설정되어있을테니 중복은 제거해주세요

ljy2855 commented on src/routers/captcha.py:
```diff
@@ -1,108 +1,56 @@
 
 @router.post("/predict")
-def predict(req: CaptchaRequest, request: Request):
+async def predict(req: CaptchaRequest, request: Request):
```
혹시 어떤 이유로, 비동기 처리로 추가하셨을까요?

ljy2855 commented on src/saved_images/captcha_55b8927b-0f94-45a2-a39e-888ad1486bf9.png:
saved images 폴더는 gitignore로 올려주심 감사하겠습니다 :)

ljy2855 commented on src/main.py:
```diff
@@ -22,4 +29,67 @@
+class InferenceRequest(BaseModel):
+    inputs: List[List[float]]
+
+@app.get("/")
```
main에 있는 endpoint를 router로 옮겨주셔요! `routers/model.y`

ljy2855 commented on src/main.py:
```diff
@@ -22,4 +29,67 @@
     max_age=3600,  # 세션 유지 시간 (초)
 )
 
+class InferenceRequest(BaseModel):
```
마찬가지로 Request model도 옮겨주셔요

ljy2855 commented on src/routers/captcha.py:
```diff
@@ -1,108 +1,56 @@
         
+        predicted_digit = result["predictions"][0]
         passed = str(predicted_digit) == expected
+        
```
혹시 `pre-commit install` 터미널에 한번 실행 가능할까요?

RaonJena99 committed:
백엔드 구조 개선

- 모델 예측 관련 endpoint를 main.py에서 router/model.py로 분리하여 구조화
- 저장된 전처리 이미지를 .gitignore 처리
